### PR TITLE
add aliasing suppression to resample function

### LINKED ISSUE #328: Artifacts in dsp.resample
## General

Resample introduces out of band artifacts when upsampling. It could be mitigated by an additional low-pass.

### PR DESCRIPTION
**Pull request by f-brinkmann**
### Which issue(s) are closed by this pull request?

Closes #328 

### Changes proposed in this pull request:

add aliasing suppression to resample function. To see this in action, e.g., use

```python
import pyfar as pf

signal = pf.signals.noise(512, seed=7)

true = pf.dsp.resample(signal, 48000, suppress_aliasing=True)
false = pf.dsp.resample(signal, 48000, suppress_aliasing=False)

pf.plot.time_freq(false, label="false")
ax = pf.plot.time_freq(true, label="true")

ax[1].legend()
```
supression

### CODE REVIEW

File: pyfar/dsp/interpolation.py [hunk 1]
```diff
@@ -495,6 +496,14 @@ def resample(signal, sampling_rate, match_amplitude="auto", frac_limit=None):
         realizing the target sampling rate.
 
         The default is ``None``, which uses ``frac_limit = 1e6``.
+    suppress_aliasing : bool, optional
+        Suppress aliased energy above the Nyquist frequency of the input
+        signal, i.e., ``signal.sampling_rate/2``. This is recommended, and is
```
Comment (sikersten) [inline pyfar/dsp/interpolation.py:501]: Default indication missing, could e.g. be inserted here:
```suggestion
        signal, i.e., ``signal.sampling_rate/2``. This is recommended, the default is ``True``. It is
```

File: pyfar/dsp/interpolation.py [hunk 2]
```diff
@@ -495,6 +496,14 @@ def resample(signal, sampling_rate, match_amplitude="auto", frac_limit=None):
...
+        realized by applying a zero-phase Elliptic filter with a pass band
+        ripple of 0.1 dB, a stop band attenuation of 60 dB, and pass/stop
+        band frequencies of ``signal.sampling_rate/2`` and
+        ``min(1, 1.05 * signal.sampling_rate/2)`` Hz. Note that this is only
```
Comment (sikersten) [inline pyfar/dsp/interpolation.py:505]: This is a little confusing.. it just says that the passband edge frequency is the old Nyquist frequency and the stopband edge frequency is the minimum of (new Nyquist frequency/1.2*old Nyquist frequency), right?

File: tests/test_dsp_resample.py [hunk 3]
```diff
@@ -128,3 +130,17 @@ def test_resample_multidimensional_impulse():
     sinc = pf.dsp.time_window(sinc, [N*L/2-int(c/2), N*L/2+int(c/2)],
                               window='hamming')
     np.testing.assert_almost_equal(resampled.time, sinc.time, decimal=2)
+
+
+def test_resample_suppress_aliasing():
```
Comment (sikersten) [inline tests/test_dsp_resample.py:135]: Maybe add test for the filter properties like zero-phase, attenuation?

**Review by sikersten — COMMENTED** [review]
Thanks for implementing, only minor comments, otherwise approved :)

File: pyfar/dsp/interpolation.py (hunk 1 above)
Comment (f-brinkmann) [reply pyfar/dsp/interpolation.py:501]: Added the default.

File: pyfar/dsp/interpolation.py (hunk 2 above)
Comment (f-brinkmann) [reply pyfar/dsp/interpolation.py]: Yes, I tried to make that more clear

File: tests/test_dsp_resample.py (hunk 3 above)
Comment (f-brinkmann) [reply tests/test_dsp_resample.py:136]: Good point, improved the testing to asses pass- and stop-band properties more precicely

**Review by sikersten — APPROVED** [review]

**Comment by mberz** [comment]
I'm not sure if this is a fair comparison. Your 'true' reference is just the result of your function. I've checked this against deterministic signals using the following:

``` python
import numpy as np
import pyfar as pf
import matplotlib.pyplot as plt

signal = pf.signals.sine(n_samples=512, frequency=1e3, sampling_rate=44100, full_period=True)
true = pf.signals.sine(n_samples=512, frequency=1e3, sampling_rate=48e3, full_period=True)

signal = pf.signals.impulse(n_samples=512, sampling_rate=44100)
true = pf.signals.impulse(n_samples=512, sampling_rate=48e3)

# true = pf.dsp.resample(signal, 48000, )
aa_filtered = pf.dsp.resample(signal, 48000)
non_aa_filtered = pf.dsp.resample(signal, 48000, suppress_aliasing=False)
pf.plot.time(aa_filtered)

plt.figure(figsize=(8, 6))
ax = pf.plot.time(true, label='ref', dB=True)
pf.plot.time(aa_filtered, linestyle='-.', label='filted', dB=True)
pf.plot.time(non_aa_filtered, linestyle='--', label='not filtered', dB=True)
ax.legend()
# ax.set_xlim(-0.001, None)
# ax.set_ylim(-20, 5)

plt.figure(figsize=(8, 6))
ax = pf.plot.freq(true, label='ref')
pf.plot.freq(aa_filtered, linestyle='-.', label='filted')
pf.plot.freq(non_aa_filtered, linestyle='--', label='not filtered')
ax.legend()
ax.set_ylim(-20, 5)

```
Which for example results in this resampled impulse:
 
output
 
The issue is the additional ripple introduced by your filter on top of an apparently existing filter I guess:

output

For a sine for example there's obviously no real influence on the time domain, as long as the frequency is far below the Nyquist frequency:

output

output

So all in all I'm a bit reluctant to merge this as the new default. 
Can you provide an example with a true reference?

**Comment by f-brinkmann** [comment]
I detected the issue for the HRTFs in #315. There you see a clear benefit of the additional filter:

```python
import pyfar as pf
import matplotlib.pyplot as plt

hrir, sources = pf.signals.files.hrirs()
hrir = hrir[0, 0]

hrir_48_aliased = pf.dsp.resample(hrir, 48e3, suppress_aliasing=False)
hrir_48_clean = pf.dsp.resample(hrir, 48e3, suppress_aliasing=True)

pf.plot.freq(hrir_48_clean, label='clean')
pf.plot.freq(hrir_48_aliased, c='y', ls="--", label='aliased')

ax.legend()
plt.savefig(r"aliasing.png")
```

aliasing

**Comment by mberz** [comment]
Ah, I see. This is an interpolation artifact, not aliasing. I think we should discuss this in the weekly.

**Comment by f-brinkmann** [comment]
We decided to change the default to ``False`` and look for a more correct and intuitive parameter name
